[WIP] Adding inorganic crystal featurizers

ncfrey opened the pull request:
### Summary
This WIP PR adds featurizers for inorganic crystal structures.
- Chemical fingerprint, Sine Coulomb Matrix, and Structure Graph featurizers

### To do
- Add methods for further featurizing structure graphs

### Additional dependencies
`deepchem.feat.materials_featurizers.py` introduces dependencies on `pymatgen` and `matminer`.

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+
+from matminer.featurizers.composition import ElementProperty
+
+from pymatgen import Composition, Structure
```
Can you move these imports into the functions/methods where they are used? DeepChem has a system for handling optional dependencies (which we need to document...). Check out https://github.com/deepchem/deepchem/blob/master/deepchem/utils/rdkit_util.py for an example of how to do this

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+class ChemicalFingerprint(Featurizer):
+  """
+  Chemical fingerprint of elemental properties from composition.
+
```
Can you expand out the docstring here to add more details about the method? This should include MLA style citations to any papers that provide inspirations, a mention of the pymatgen depdency for this featurizer and a general discussion of when this would be useful.

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+  """
+  Chemical fingerprint of elemental properties from composition.
+
+  Parameters
```
Can you move the `Parameters` section down to the docstring for `__init__`?

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+  """
+  Calculate sine Coulomb matrix for crystals.
+
+  Variant of Coulomb matrix for periodic crystals
```
Can you expand out this docstring as mentioned above? Usually a good paragraph or two is the right length for documentation for featurizers

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+class StructureGraphFeaturizer(Featurizer):
+  """
+  Calculate structure graph for crystals.
+
```
Can you add a good docstring here as well?

rbharath commented on deepchem/feat/tests/test_materials_featurizers.py:
```diff
@@ -0,0 +1,66 @@
+    """
+
+    featurizer = ChemicalFingerprint(data_source='matminer')
+
```
A couple of good tests to add here might be the following:

1) A shape test, that checks the featurizer produces outputs of the right shape
2) A output sanity check, that checks that the output satisfies some sensible properties for this featurization

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,178 @@
+from pymatgen.analysis.local_env import MinimumDistanceNN
+
+
+class ChemicalFingerprint(Featurizer):
```
Can you add documentation listings for these new featurizers in https://github.com/deepchem/deepchem/blob/master/docs/featurizers.rst?

If we do this, then the docs will automatically render the docstrings and API for these classes.

ncfrey commented:
> Thanks for separating this out! This feature looks really exciting :) I've done a first review pass over the code. I'm really excited to see these featurizers get into DeepChem.
> 
> A general comment about documentation is that it's useful to write a good section of text. Imagine the sort of text you might add in the methods section of a paper introducing the technique and that's probably close to the right amount needed.

Thanks so much for the code review! I will work on expanding the documentation and address all your other comments!

rbharath commented:
Did a quick skim and changes are looking good! Let me know when you'd like a second review :)

ncfrey commented:
> Did a quick skim and changes are looking good! Let me know when you'd like a second review :)

Awesome, yeah I think I'm ready for another review!

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+    ----------
+    comp : str
+      Reduced formula of crystal.
+
```
Can you add a returns section to the docstring? Something like

```
Returns
--------
feats: np.ndarray
  TODO ADD DESCRIPTION
```

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+  based on elemental stoichiometry. E.g., the average electronegativity
+  of atoms in a crystal structure. The chemical fingerprint is a 
+  vector of these statistics. For a full list of properties and statistics,
+  see ElementProperty(data_source).feature_labels().
```
Can you add a reference that you mean from `matminer.featurizers.composition import ElementProperty` and put the code in brackets for correct rendering?

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+    ----------
+    struct : dict
+      pymatgen structure dictionary
+
```
Returns section for docstring

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+    Parameters
+    ----------
+    s : pymatgen structure
+
```
Returns section for docstring

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+
+    Parameters
+    ----------
+    s : pymatgen structure
```
Can you add a description for this parameter?

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+    ----------
+    struct : dict
+      pymatgen structure dictionary.
+
```
returns section

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+import numpy as np
+
+from deepchem.feat import Featurizer
+from deepchem.utils import pad_array
```
Can you update the README.md file at the top of the repo adding `pymatgen` and `matminer` as optional dependencies and installation directions? I think you'll also need to update the `.travis.yml` file so the test environment installs these as well

rbharath commented on docs/featurizers.rst:
```diff
@@ -116,6 +116,23 @@ AtomConvFeaturizer
 .. autoclass:: deepchem.feat.NeighborListComplexAtomicCoordinates
   :members:
 
+ChemicalFingerprint
```
Can you add a section for "Materials Featurizers" that all of these featurizers sits in and add some more documentation (say a paragraph) explaining how materials featurizers differ from the other featurizers and the types of problems they should be applied on?

rbharath replied:
This looks great, thanks!

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+    Parameters
+    ----------
+    struct : dict
+      pymatgen structure dictionary
```
Can you add more details on what a `pymatgen structure dictionary` is? Perhaps a reference to some of the pymatgen documentation specifying what it is would be useful.

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,250 @@
+      return sine_mat
+
+
+class StructureGraphFeaturizer(Featurizer):
```
Can you add a unit test for `StructureGraphFeaturizer`?

ncfrey commented:
> Looking much better! I've done a second pass mainly requesting more docs and more tests but we're getting closer for sure :).

Great, thanks, ready to review! 

I added more to the docstrings and added a unittest for the StructureGraphFeaturizer. I also added `pymatgen` and `matminer` to the `install_deepchem_conda.sh` script, which I think should take care of the CI build, but I'm not sure. Is there a specific place where I should add details on how to install those packages? I didn't see any instructions for other soft dependencies.

rbharath commented on scripts/install_deepchem_conda.sh:
```diff
@@ -33,7 +33,8 @@ conda install -y -q -c deepchem -c rdkit -c conda-forge -c omnia \
     pytest-cov \
     flaky
-yes | pip install -U tensorflow tensorflow-probability
+yes | pip install -U matminer tensorflow tensorflow-probability
```
Can you also add this to `scripts/install_deepchem_conda.ps1`?

ncfrey replied:
Will do!

nd-02110114 commented:
I will review until tomorrow!

nd-02110114 commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,285 @@
+
+    atom_features = np.array([site.specie.Z for site in struct], dtype='int32')
+
+    sg = StructureGraph.with_local_env_strategy(struct, self.strategy)
```
From my experience, `with_local_env_strategy ` is too slow... (This issue is discussed in CGCNN reop, see https://github.com/txie-93/cgcnn/pull/18) 
If this featurizer is implemented for GCN or CGCNN, I recommend using `get_all_neighbors ` method like https://github.com/txie-93/cgcnn/blob/d612a69530a72ba686fca56813657b89f2440cc5/cgcnn/data.py#L319-L350

ncfrey replied:
Agreed, I think doing something like `get_all_neighbors` is a great idea. I frequently have to transform a bunch of structures into structure graphs and save them and it does take quite awhile.

rbharath replied:
+1 for using `get_all_neighbors` here instead.

ncfrey commented:
> I did review and commented!
> 
> @rbharath
> To be honest, I think `Chemical fingerprint` and `Sine Coulomb Matrix` are not necessary for DeepChem. This is because these featurizers are too thin wrappers of matminer and these features cannot be used with many DeepChem models. So, it is better for DeepChem to provide these codes as as a tutorial for making user-defined featurizers.
> 
> On the other hand, I seem that `Structure Graph featurizers` is helpful for DeepChem. This is because this feature has possiblitiy to be used with some DeppChem models like GCN, MPNN. (I seem it is possible to add custom default_generator method for crystal graphs about each Model) So, I think this PR should focus on `Structure Graph Featurizers`.

Hey @nd-02110114, thanks for the review! I will leave it up to you and @rbharath if you think `Chemical fingerprint` and `SCM` are good additions or not. They are both standard representations for crystal structures and they will work with any DeepChem model that takes vector inputs, but they are both basically wrappers for matminer featurizers.

ncfrey commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,285 @@
+
+    return features
+
+  def sine_coulomb_matrix(self, s):
```
Here I didn't use matminer to limit the number of methods that require matminer as a soft dependency, and because the implementation of SCM is nearly identical to the one that is already in `dc.feat.coulomb_matrices`. I can certainly use the matminer module if that would be better.

nd-02110114 replied:
If matminer's codes have some mistakes and they are modified,  it is difficult to sync the modification.
So, i think it is better to use matminer's module directly.  

On the other hand, I think you know more deeply about this descriptor. If you seem this implementation is exactly correct, I think it is not necessary to use matminer's module directly. 

To be honest, I couldn't judge whether this implementation is definitely correct or not....  I confirmed the definition of SineCoulombMatrix from eq. 23 and 24 of [this thesis](https://arxiv.org/pdf/1503.07406.pdf), but I seem this implementation is a little tricky... (Especieally, I seem `trig_dist` calculation is a little strange... eq.24 includes the inverse matrix calculation, but such a caluculation is not seen in this implementation)

rbharath replied:
Really good points in this thread. Doing some thinking, because of the similarity of `SCM` to `dc.feat.coulomb_matrices`, perhaps the right way to handle this is to expand the existing `CoulombMatrix` Featurizer to accept JSON crystal structures and use sine as the distance function.

@ncfrey Would it be straightforward to merge the SCM code into the existing Coulomb featurizer?

+1 for the idea of reusing the `matminer` implementation if possible if it works for our purposes. Less code to maintain makes life easy :)

ncfrey replied:
I think that if `Featurizer` and `CoulombMatrix` Featurizer are updated to accept Json files, then it should be no problem to generalize to SCM. For now, I'll move to the `matminer` implementation to keep things simple and easy to maintain.

rbharath replied:
Makes sense! Perhaps we can revisit this in a future PR once the JSON file handling infrastructure is in place.

nd-02110114 commented:
@ncfrey CC. @rbharath 
Thank you for contributing!  My research topic is related to inorganic material design and I also wanted to start inorganic crystal supports for DeepChem.

After I rethink whether Chemical fingerprint and SCM are good additions or not, I think it is not problem to add them.  I'm sorry to say different opinion than before :bowing_man:

> these features cannot be used with many DeepChem models. 

It is not correct. These features can be used with sklean_models and xgboost_models.


> these featurizers are too thin wrappers of matminer

Some featurizers like `CircularFingerprint` was also a thin wrapper of RDKit.

ncfrey commented:
> @ncfrey CC. @rbharath
> Thank you for contributing! My research topic is related to inorganic material design and I also wanted to start inorganic crystal supports for DeepChem.
> 
> After I rethink whether Chemical fingerprint and SCM are good additions or not, I think it is not problem to add them. I'm sorry to say different opinion than before 🙇‍♂️
> 
> > these features cannot be used with many DeepChem models.
> 
> It is not correct. These features can be used with sklean_models and xgboost_models.
> 
> > these featurizers are too thin wrappers of matminer
> 
> Some featurizers like `CircularFingerprint` was also a thin wrapper of RDKit.

That's awesome! My goal with my initial contributions is to begin to add some inorganic crystal support while I learn about DeepChem, and then get more involved in the cheminformatics and biology side because I don't have any experience in that area. But I would be happy to help out if you are going to work more on supporしますor crystals.

初めまして。ナタンです。よろしくお願いします。日本語が少し話します。

rbharath commented:
Awesome discussion here folks! I took a look through and my two cents is that ChemicalFingerprint and SCM would be useful featurizers for DeepChem. Vectorial featurizers are limited in the models they apply to (Sklearn, XGBoost, Fully connected networks) but are really powerful as baselines before starting work on a more complex method. The `dc.feat.CircularFingerprint` for example is a thin wrapper over RDKit but I think a lot of folks (including me) find it very useful.

I'll do a pass on the various threads above shortly

rbharath commented on deepchem/feat/materials_featurizers.py:
```diff
@@ -0,0 +1,285 @@
+from deepchem.utils import pad_array
+
+
+class ChemicalFingerprint(Featurizer):
```
@nd-02110114 Great point about the need to modify `Featurizer` I have an upstream PR (part of #1792) that does this. I'll try to get this merged in sooner :)

+1 for the idea of `ElementPropertyFingerprint` instead of `ChemicalFingerprint`.

ncfrey commented:
I've changed the `SineCoulombMatrix` featurizer to make use of the `matminer` implementation, and also altered `StructureGraphFeaturizer` to use the `get_all_neighbors` method as in CGCNN.

rbharath commented:
I think at this point all outstanding review questions have been addressed. Is this PR ready to merge in? @nd-02110114 @peastman @ncfrey

ncfrey commented:
> I think at this point all outstanding review questions have been addressed. Is this PR ready to merge in? @nd-02110114 @peastman @ncfrey

Happy to make any further changes if needed, but otherwise good to go from me!

nd-02110114 commented:
I think this PR is ready to merge! 
But we should care about this point.

>  these featurizer can not be used with CSVLoader because CSVLoader execute the conversion from SMILES to mol object internally. So, I seem we should determine custom Dataloader design for inorganic crystal. (It is OK to split PR about this)

ncfrey commented:
> I think this PR is ready to merge!
> But we should care about this point.
> 
> > these featurizer can not be used with CSVLoader because CSVLoader execute the conversion from SMILES to mol object internally. So, I seem we should determine custom Dataloader design for inorganic crystal. (It is OK to split PR about this)

Definitely! I previously wrote separate dataloader classes to load crystal structures. I think it makes sense to have a new `JsonLoader` class, but I am now working on generalizing `Featurizer` so that `CSVLoader` can be used with crystal inputs.

rbharath commented:
@ncfrey Adding on the discussions, in my giant running PR (#1792) I started some work on overhauling the featurizers. The idea was that `Featurizer` become a totally abstract class for any type of data and I added a new `MolecularFeaturizer` class for molecules that became the superclass of the molecular featurizers. See https://github.com/deepchem/deepchem/blob/f07b05185b1e9a84846f1e1b8dc50ea1d8af97c6/deepchem/feat/base_classes.py#L14

Would you be able to incorporate these changes into your current work? I'm unfortunately behind on these changes due to my other PRs, so it would be really useful to get the abstract `Featurizer` and `MolecularFeaturizer` added. This suggests that perhaps we should add another abstract `CrystalFeaturizer` class that is the parent class to all crystal featurizers.

rbharath commented:
To clarify, this would be in a different PR than this one of course! I think this one is ready to merge as is. I'll merge on Monday in case there are any other comments

ncfrey commented:
> @ncfrey Adding on the discussions, in my giant running PR (#1792) I started some work on overhauling the featurizers. The idea was that `Featurizer` become a totally abstract class for any type of data and I added a new `MolecularFeaturizer` class for molecules that became the superclass of the molecular featurizers. See
> 
> https://github.com/deepchem/deepchem/blob/f07b05185b1e9a84846f1e1b8dc50ea1d8af97c6/deepchem/feat/base_classes.py#L14
> 
> Would you be able to incorporate these changes into your current work? I'm unfortunately behind on these changes due to my other PRs, so it would be really useful to get the abstract `Featurizer` and `MolecularFeaturizer` added. This suggests that perhaps we should add another abstract `CrystalFeaturizer` class that is the parent class to all crystal featurizers.

Sounds good! I will add these into my featurizer work and implement a `CrystalFeaturizer.`

rbharath commented:
@ncfrey Ah looks like a merge conflict has arisen! We'll need to fix this before we merge in. Can you rebase this PR? In case you haven't done this before, here's the set of git commands you'll need to make (starting from the branch for this PR)

```
# Checkout master
git checkout master
# Fetch updates. Assumes the remote is called origin
git fetch origin
# Pull updates into master branch
git pull
# Swap back to materials_featurizers branch
git checkout materials_featurizers
# Rebase, that is put changes in this PR on top of latest changes
git rebase origin/master
```
While rebasing you'll run into the conflict files. When the conflicts arise, open the files that have conflicts in an editor then manually fix them. Then add them and continue the rebase
```
# You only need to add the files that have conflicts at this stage in the rebase, but for simplicity I'm assuming both have conflicts
git add scripts/install_deepchem_conda.ps1
git add scripts/install_deepchem_conda.sh
# Continue rebase
git rebase --continue
```
You might have to repeat this last set of instructions a few times if there are multiple conflicts that need to be fixed

ncfrey commented:
> @ncfrey Ah looks like a merge conflict has arisen! We'll need to fix this before we merge in. Can you rebase this PR? In case you haven't done this before, here's the set of git commands you'll need to make (starting from the branch for this PR)

It looks like the conflicts are only in the version specifications for tensorflow in the install scripts, so I may be able to manually resolve them without rebasing? I'm not sure - I can try that first and rebase if necessary.
